src/webgpu/shader/validation/expression/call/builtin/textureSampleCompare.spec.ts

```diff
@@ -0,0 +1,250 @@
+
+type TextureSampleCompareArguments = {
+  coordsArgType: ScalarType | VectorType;
+  hasArrayIndexArg?: boolean;
+  offsetArgType?: VectorType;
+};
+
+const kValidTextureSampleCompareParameterTypes: { [n: string]: TextureSampleCompareArguments } = {
+  texture_depth_2d: { coordsArgType: Type.vec2f, offsetArgType: Type.vec2i },
+  texture_depth_2d_array: {
+    coordsArgType: Type.vec2f,
+    hasArrayIndexArg: true,
+    offsetArgType: Type.vec2i,
+  },
+  texture_depth_cube: { coordsArgType: Type.vec3f },
+  texture_depth_cube_array: { coordsArgType: Type.vec3f, hasArrayIndexArg: true },
+} as const;
+
+const kTextureTypes = keysOf(kValidTextureSampleCompareParameterTypes);
+const kValuesTypes = objectsToRecord(kAllScalarsAndVectors);
+
+export const g = makeTestGroup(ShaderValidationTest);
+
+g.test('coords_argument')
+  .desc(
+    `
+Validates that only incorrect coords arguments are rejected by ${builtin}
+`
+  )
+  .params(u =>
+    u
+      .combine('textureType', keysOf(kValidTextureSampleCompareParameterTypes))
+      .combine('coordType', keysOf(kValuesTypes))
+      .beginSubcases()
+      .combine('value', [-1, 0, 1] as const)
+      .expand('offset', ({ textureType }) => {
+        const offset = kValidTextureSampleCompareParameterTypes[textureType].offsetArgType;
+        return offset ? [false, true] : [false];
+      })
+  )
+  .fn(t => {
+    const { textureType, coordType, offset, value } = t.params;
+    const coordArgType = kValuesTypes[coordType];
+    const {
+      offsetArgType,
+      coordsArgType: coordsRequiredType,
+      hasArrayIndexArg,
+    } = kValidTextureSampleCompareParameterTypes[textureType];
+
+    const coordWGSL = coordArgType.create(value).wgsl();
+    const arrayWGSL = hasArrayIndexArg ? ', 0' : '';
+    const offsetWGSL = offset ? `, ${offsetArgType?.create(0).wgsl()}` : '';
+
+    const code = `
+@group(0) @binding(0) var s: sampler_comparison;
+@group(0) @binding(1) var t: ${textureType};
+@fragment fn fs() -> @location(0) vec4f {
+  let v = textureSampleCompare(t, s, ${coordWGSL}${arrayWGSL}, 0${offsetWGSL});
+  return vec4f(0);
+}
+`;
+    const expectSuccess = isConvertible(coordArgType, coordsRequiredType);
+    t.expectCompileResult(expectSuccess, code);
+  });
+
+g.test('array_index_argument')
+  .desc(
+    `
+Validates that only incorrect array_index arguments are rejected by ${builtin}
+`
+  )
+  .params(u =>
+    u
+      .combine('textureType', kTextureTypes)
+      // filter out types with no array_index
+      .filter(
+        ({ textureType }) =>
+          !!kValidTextureSampleCompareParameterTypes[textureType].hasArrayIndexArg
+      )
+      .combine('arrayIndexType', keysOf(kValuesTypes))
+      .beginSubcases()
+      .combine('value', [-9, -8, 0, 7, 8])
+  )
+  .fn(t => {
+    const { textureType, arrayIndexType, value } = t.params;
+    const arrayIndexArgType = kValuesTypes[arrayIndexType];
+    const args = [arrayIndexArgType.create(value)];
```

Comment:
It took me a while to figure out why this was not causing failures for `u32`. `value` includes `-9`, `-8`, and I was expecting this to result in `-9u`, `-8u` which would result in: `error: no matching overload for operator - (u32)`.

However, `ScalarType.create()` calls `u32(value)`, which is implemented as:
https://github.com/gpuweb/cts/blob/f1aa9428a23f53c98aaa9588af36e319b37ed94e/src/webgpu/util/conversion.ts#L1631-L1635

So, we end up with the two's complement. This is unexpected in this context.

We're not trying to test for unsigned integers handling negative numbers, so I'd suggest changing the `value` range to just include positive numbers.

Same thing applies for the other tests.

Reply:
I see what you're saying about `-8u` getting converted via `Uint32Array` but I need negative numbers to test `offset` which is required to generate an error for values < -8 and it will end up doing the same thing, passing `-8` to `offsetArgType.create()` with those values when `offsetArgType` is `u32`.

I switched to filtering out negative values with unsigned types